### ISSUE TITLE
torch.nn.TransformerEncoder hits error with thunder.jit

### ISSUE THREAD

**Issue by Fuzzkatt** [Lightning-AI/lightning-thunder#289]
Tracking my attempt to make torch.nn.Transformer work with thunder.jit. Currently after applying a series of workarounds, we can get torch.nn.TransformerEncoderLayer to work with thunder.jit. However, we hit a new error with torch.nn.TransformerEncoder:

`transformer: thunder.core.interpreter.InterpreterError: Encountered exception AssertionError: attribute size of TensorProxy object out of sync: <function prop_lookaside_wrap.<locals>.fn.<locals>.fn_ at 0x7fe1c457a440> vs. <function prop_lookaside_wrap.<locals>.fn.<locals>.fn_ at 0x7fe1c457b0a0> while tracing Transformer(`

which mostly just stacks n copies of TransformerEncoderLayers into one module. 

Workarounds: 
* comment out is_nested check in https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/activation.py#L1236-L1238
* rewrite unflatten as reshape in https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L4909 (Ex. https://github.com/Lightning-AI/lightning-thunder/issues/288) 
* comment out is_nested check in https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L42

Minimal repro:

```
import torch
import thunder
from thunder.examine import examine

src = torch.randn(10, 32, 512, device='cuda')
encoder_layer = torch.nn.TransformerEncoderLayer(d_model=512, nhead=8).to('cuda')
encoder = torch.nn.TransformerEncoder(encoder_layer, num_layers=6).to('cuda')

# works after applying workarounds
jfunc = thunder.jit(encoder_layer)
jfunc(src)
print('encoder layer working')

# fails even after applying workarounds
jfunc = thunder.jit(encoder)
jfunc(src)

```

**Comment by Fuzzkatt** [comment]
https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L921 using copy.deep_copy which is called here https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L276 looks suspicious; how would this work with thunder.jit?

**Comment by mruberry** [comment]
triage review — @t-vi can add more details about this issue

**Comment by Fuzzkatt** [comment]
Narrowed down issue to much more minimal issue here: https://github.com/Lightning-AI/lightning-thunder/issues/349. In this particular case, the two size() checks happen first in https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L45 as the forward of TransformerEncoder calls _get_seq_len as part of the is_causal check here https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L411. The second call is from https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L4903 which is from the forward of TransformerEncoder calling the stack of TransformerEncoderLayer in https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L415.

**Comment by Fuzzkatt** [comment]
Found possible fix https://github.com/Lightning-AI/lightning-thunder/pull/352. @mruberry @t-vi could you take a look and comment on whether the fix is acceptable or give any suggestions?

**Comment by Fuzzkatt** [comment]
Fixed with #352, closing
